code-generation/llms-vscode-integration.md

```diff
@@ -0,0 +1,85 @@
+{
+  "title": "YourTitleHere",
+  "model": "YourModelName",
+  "completionOptions": {},
+  "apiBase": "http://localhost:8001/v1/",
+  "provider": "openai"
+}
+```
+
+## Step 4: Examples of Interaction
+
+Now that you've configured the "Continue" extension, let's explore how you can effectively interact with the language model directly within VS Code. Here are several ways to engage with the extension:
+
+1. **Prompting for code generation:** Open the "Continue" panel in VS Code and prompt the extension with a specific task, such as "Write a code to add two numbers." The extension will then provide relevant code autocompletion suggestions based on your input prompt, aiding in code generation and text completion tasks.
+
+   ![Prompt-response](../assets/interaction-vscode1.png)
+
+2. **Querying Working Code:** Copy your existing code snippet or press `⌘ + L` to paste it into the "Continue" panel, then pose a question such as "Explain this section of the code." The extension (LLM) will analyze the code snippet and provide explanations or insights to help you understand it better.
+
+   ![Querying Working Code](../assets/interaction-vscode2.png)
+
+3. **Editing Code in Script:** Edit your Python code directly within a `.py` script file using the "Continue" extension. Press `⌘ + I` to initiate the edit mode. You can then refine a specific line of code or request enhancements to make it more efficient. The extension will suggest additional code by replacing your edited code and provide options for you to accept or reject the proposed changes.
+
+   ![Editing Code in Script](../assets/interaction-vscode3.png)
+
+By exploring these interactions, users can fully leverage the capabilities of language models within VS Code, enhancing their coding experience and productivity.
+
+4. **Tab Autocomplete:**
+
+![autocompletion-config-example](../assets/autocompletion-example.png)
+
+ The "Continue" extension also offers a tab autocomplete feature in its pre-release version. To set up autocomplete with a custom model, you'll need to edit the `config.json` file as shown below:
+
+```
+{
+    "tabAutocompleteModel": {
+        "title": "Tab Autocomplete Model",
+        "provider": "provider name",
+        "model": "model name",
+        "apiBase": "https://<endpoint>"
+    },
+    ...
+}
+```
+
+The tabAutocompleteModel is similar to other objects in the models array of config.json. You have the flexibility to choose any model you prefer, but it's recommended to use a small model for tab autocomplete, such as deepseek-1b, starcoder-1b,  starcoder-3b, or stable-code-3b for optimal performance.
+
+To obtain the endpoint for this model, you can utilize llama-cpp-python, which provides an OpenAI API-compatible server. This web server serves local models and facilitates easy connections with existing clients.
+
+Here's how you can set it up:
+
+```
+pip install llama-cpp-python[server]
+
+# Start the server
+python3 -m llama_cpp.server --model <model_path>
+``` 
+
```

Comment:
Why are you suggesting to do it this way instead of using the llamacpp playground container we use in the rest of the repo?

Reply:
Here we have two different cases, 
1. where we interact with models using prompt, edit and code explanation features. 
2. where we interact with autocomplete feature, which require different model. 

In this article, I have suggested another way to setup server for model used for autocompletion. If anyone wants to interact with both features, then they can used llama-cpp playground container to interact with case 1, and setup llama-cpp-python server for case 2. Does this approach makes sense to you? wdyt?
Thanks.

Reply:
Thats a good question.

- yes one can create a second playground container with the other model.

- No there is not such restriction. One can use same model for both cases. However, they strongly recommend to use smaller model (1B - 3B parameters) for TabAutoCompletion options. Currently, I am exploring some models which I can use in both cases, without frequent crashes.

Reply:
Yes, this is what I experienced while using containerized options, frequent crashing. I hesitate to suggest 2 models, or a non-containerized approach or one that crashes frequently. It would be ideal if we can figure out why its crashing and come up with a solution to use 1 containerized model server for both.

Reply:
Hi @MichaelClifford ,

I observed that the reason for the crashes is because we're sending too many requests to our model too quickly, especially when we're using the AutoComplete feature. To fix this, I increased the `debounceDelay` parameter  setting to 4000 ms so that there is a short wait time between each request we send to the server. Now it seems to be working. ptal. 

Thank you!